### PR TITLE
A large set of improvements to the Stackdriver components.

### PR DESCRIPTION
**Pull request by bmoyles0117**
**What this PR does / why we need it**:
This PR delivers a large set of improvements for both the Stackdriver Logging agent and the Stackdriver Metadata agent.

**Release note**:
```release-note
Metadata Agent Improvements
Bump metadata agent version to 0.2-0.0.21-1.
Expand the metadata agent's access to all API groups.
Remove metadata agent config maps in favor of command line flags.
Update the metadata agent's liveness probe to a new /healthz handler.

Logging Agent Improvements
Bump logging agent version to 0.2-1.5.33-1-k8s-1.
Appropriately set log severity for k8s_container.
Fix detect exceptions plugin to analyze message field instead of log field.
Fix detect exceptions plugin to analyze streams based on local resource id.
Disable the metadata agent for monitored resource construction in logging.
Disable timestamp adjustment in logs to optimize performance.
Reduce logging agent buffer chunk limit to 512k to optimize performance.
```

### CODE REVIEW

**Comment by igorpeshansky** [comment]
See #66091.

File: cluster/addons/fluentd-gcp/fluentd-gcp-configmap.yaml [hunk 1]
```diff
@@ -421,8 +423,7 @@ data:
       # Use multiple threads for processing.
       num_threads 2
       use_grpc true
-      # Use Metadata Agent to get monitored resource.
-      enable_metadata_agent true
+      adjust_invalid_timestamps false
```
Comment (qingling128) [inline cluster/addons/fluentd-gcp/fluentd-gcp-configmap.yaml]: Some comment?

```
      # Skip timestamp adjustment as this is in a controlled environment with
      # known timestamp format. This helps with CPU usage.
```

Comment (qingling128) [reply cluster/addons/fluentd-gcp/fluentd-gcp-configmap.yaml]: We should apply this to `fluentd-gcp-configmap-old.yaml` as well.

File: cluster/addons/fluentd-gcp/fluentd-gcp-configmap.yaml [hunk 2]
```diff
@@ -408,7 +410,7 @@ data:
       buffer_queue_full_action block
       # Set the chunk limit conservatively to avoid exceeding the recommended
       # chunk size of 5MB per write request.
-      buffer_chunk_limit 1M
+      buffer_chunk_limit 512k
```
Comment (qingling128) [inline cluster/addons/fluentd-gcp/fluentd-gcp-configmap.yaml]: We should apply this to `fluentd-gcp-configmap-old.yaml` as well.

File: cluster/addons/fluentd-gcp/fluentd-gcp-configmap.yaml [hunk 3]
```diff
@@ -108,8 +110,8 @@ data:
       @type detect_exceptions
 
       remove_tag_prefix raw
-      message log
-      stream stream
+      message message
```
Comment (qingling128) [inline cluster/addons/fluentd-gcp/fluentd-gcp-configmap.yaml]: Seems GitHub is not showing the latest `kubernetes:master` code. `message log` has already been [fixed](https://github.com/kubernetes/kubernetes/blob/master/cluster/addons/fluentd-gcp/fluentd-gcp-configmap.yaml#L111) in https://github.com/kubernetes/kubernetes/commit/8f340c6c6a137b968fd7fc2e730c7dcdf667d624.

We might need to rebase `bmoyles0117:apply-latest-stackdriver-fixes`.

File: cluster/gce/gci/configure-helper.sh [hunk 4]
```diff
@@ -2184,7 +2184,7 @@ function setup-fluentd {
     fluentd_gcp_configmap_name="fluentd-gcp-config-old"
   fi
   sed -i -e "s@{{ fluentd_gcp_configmap_name }}@${fluentd_gcp_configmap_name}@g" "${fluentd_gcp_yaml}"
-  fluentd_gcp_version="${FLUENTD_GCP_VERSION:-0.2-1.5.30-1-k8s}"
+  fluentd_gcp_version="${FLUENTD_GCP_VERSION:-0.2-1.5.33-1-k8s-1}"
```
Comment (qingling128) [inline cluster/gce/gci/configure-helper.sh]: https://github.com/GoogleCloudPlatform/google-fluentd/pull/100 is pinning `systemd-journal` version to avoid crash loops due to a typo in their latest release. We need to update this to `0.2-1.5.33-1-k8s-2` once that change is soaked and verified.

**Review by qingling128 — CHANGES_REQUESTED** [review]

File: cluster/addons/fluentd-gcp/fluentd-gcp-configmap.yaml (hunk 2 above)
Comment (bmoyles0117) [reply cluster/addons/fluentd-gcp/fluentd-gcp-configmap.yaml]: I'm not sure that we want to make changes to the existing configmap, why do you want to bundle changes to existing configs?

File: cluster/addons/fluentd-gcp/fluentd-gcp-configmap.yaml (hunk 1 above)
Comment (bmoyles0117) [reply cluster/addons/fluentd-gcp/fluentd-gcp-configmap.yaml]: Applied the comment only to the new configs, same question for old configs as mentioned above.

File: cluster/gce/gci/configure-helper.sh (hunk 4 above)
Comment (bmoyles0117) [reply cluster/gce/gci/configure-helper.sh]: Done

File: cluster/addons/fluentd-gcp/fluentd-gcp-configmap.yaml (hunk 1 above)
Comment (qingling128) [reply cluster/addons/fluentd-gcp/fluentd-gcp-configmap.yaml]: Sync'ed offline. Let's leave the old config as-is for now.

File: cluster/addons/fluentd-gcp/fluentd-gcp-configmap.yaml [hunk 5]
```diff
@@ -408,7 +410,7 @@ data:
       buffer_queue_full_action block
       # Set the chunk limit conservatively to avoid exceeding the recommended
       # chunk size of 5MB per write request.
-      buffer_chunk_limit 1M
+      buffer_chunk_limit 512k
       # Cap the combined memory usage of this buffer and the one below to
       # 1MiB/chunk * (6 + 2) chunks = 8 MiB
```
Comment (qingling128) [inline cluster/addons/fluentd-gcp/fluentd-gcp-configmap.yaml]: Adjust the comment? It should be
```
512KiB/chunk * (6 + 2) chunks = 84 MiB
```
now, right?

File: cluster/addons/fluentd-gcp/fluentd-gcp-configmap.yaml (hunk 2 above)
Comment (qingling128) [reply cluster/addons/fluentd-gcp/fluentd-gcp-configmap.yaml]: Sync'ed offline. Will leave the old config as-is.

File: cluster/addons/metadata-agent/stackdriver/metadata-agent.yaml [hunk 6]
```diff
@@ -110,27 +85,22 @@ spec:
...
-          exec:
-            command:
-              - /bin/bash
-              - -c
-              - |
-                if [[ -f /var/run/metadata-agent/health/unhealthy ]]; then
-                  exit 1;
-                fi
-          periodSeconds: 10
+          httpGet:
```
Comment (qingling128) [inline cluster/addons/metadata-agent/stackdriver/metadata-agent.yaml]: Just double checking - Is this change on purpose? Or something got in by accident when addressing the rebasing conflicts?

File: cluster/addons/metadata-agent/stackdriver/metadata-agent.yaml [hunk 7]
```diff
@@ -45,27 +29,22 @@ spec:
...
```
Comment (qingling128) [inline cluster/addons/metadata-agent/stackdriver/metadata-agent.yaml]: Same here. Is this health probe change on purpose? Or something got in by accident when addressing the rebasing conflicts?

**Review by qingling128 — COMMENTED** [review]
Double checking some changes that seem new.

File: cluster/addons/fluentd-gcp/fluentd-gcp-configmap.yaml (hunk 5 above)
Comment (bmoyles0117) [reply cluster/addons/fluentd-gcp/fluentd-gcp-configmap.yaml]: Updated

File: cluster/addons/metadata-agent/stackdriver/metadata-agent.yaml (hunk 7 above)
Comment (bmoyles0117) [reply cluster/addons/metadata-agent/stackdriver/metadata-agent.yaml]: The new health checks use http-based health checks to a handler that the API server exposes. Is this your concern? I feel that I might be missing something when you point out a rebase conflict.

**Review by bmoyles0117 — COMMENTED** [review]
PTAL

File: cluster/addons/metadata-agent/stackdriver/metadata-agent.yaml (hunk 7 above)
Comment (qingling128) [reply cluster/addons/metadata-agent/stackdriver/metadata-agent.yaml]: Ah. These changes were probably there before the rebase then. I was not sure of that and wanna double check.

Looks good.

File: cluster/addons/metadata-agent/stackdriver/metadata-agent.yaml (hunk 6 above)
Comment (qingling128) [reply cluster/addons/metadata-agent/stackdriver/metadata-agent.yaml]: Same. LGTM.

File: cluster/gce/gci/configure-helper.sh [hunk 8]
```diff
@@ -2237,7 +2237,7 @@ function setup-fluentd {
     fluentd_gcp_configmap_name="fluentd-gcp-config-old"
   fi
   sed -i -e "s@{{ fluentd_gcp_configmap_name }}@${fluentd_gcp_configmap_name}@g" "${fluentd_gcp_yaml}"
-  fluentd_gcp_version="${FLUENTD_GCP_VERSION:-0.2-1.5.30-1-k8s}"
+  fluentd_gcp_version="${FLUENTD_GCP_VERSION:-0.2-1.5.33-1-k8s-2}"
```
Comment (qingling128) [inline cluster/gce/gci/configure-helper.sh]: Just double checked. Seems `0.2-1.5.33-1-k8s-2` does not exist. The latest version in the regression cluster is `0.3-1.5.34-1-k8s-1`.

Comment (bmoyles0117) [reply cluster/gce/gci/configure-helper.sh]: Done

**Review by qingling128 — APPROVED** [review]
LGTM.

**Comment by bmoyles0117** [comment]
@serathius @loburm can one of you give us ok-to-test?

File: cluster/addons/fluentd-gcp/fluentd-gcp-configmap.yaml [hunk 9]
```diff
@@ -450,15 +453,16 @@ data:
...
-      buffer_chunk_limit 1M
+      buffer_chunk_limit 512k
       buffer_queue_limit 2
       flush_interval 5s
       max_retry_wait 30
       disable_retry_limit
       num_threads 2
       use_grpc true
-      # Use Metadata Agent to get monitored resource.
-      enable_metadata_agent true
```
Comment (igorpeshansky) [inline cluster/addons/fluentd-gcp/fluentd-gcp-configmap.yaml]: Why are we turning this off? The logging agent will construct the resources itself anyway, but the metadata agent can serve as a fallback for the resources it doesn't know about.

**Review by igorpeshansky — CHANGES_REQUESTED** [review]

File: cluster/addons/fluentd-gcp/fluentd-gcp-configmap.yaml (hunk 9 above)
Comment (bmoyles0117) [reply cluster/addons/fluentd-gcp/fluentd-gcp-configmap.yaml]: Performance. Having this enabled currently results in an extreme performance degradation as opposed to doing everything in-process. This helps get us closer to exceeding the current throughput numbers achievable by the gke_container logging agent. Until we implement performance optimizations, this is a better path forward for our users, as all of the kubernetes resource types are currently fully supported.

Comment (igorpeshansky) [reply cluster/addons/fluentd-gcp/fluentd-gcp-configmap.yaml]: Fine. Let's re-enable this when those performance improvements are implemented.

**Review by igorpeshansky — APPROVED** [review]
LGTM :shipit:

File: cluster/addons/fluentd-gcp/fluentd-gcp-configmap.yaml (hunk 9 above)
Comment (loburm) [reply cluster/addons/fluentd-gcp/fluentd-gcp-configmap.yaml]: Could you give a short list of resources, that will stop working with this change?

File: cluster/addons/metadata-agent/stackdriver/metadata-agent-rbac.yaml [hunk 10]
```diff
@@ -7,9 +7,7 @@ metadata:
     addonmanager.kubernetes.io/mode: Reconcile
 rules:
 - apiGroups:
-  - ""
-  - "apps"
-  - "extensions"
+  - "*"
```
Comment (loburm) [inline cluster/addons/metadata-agent/stackdriver/metadata-agent-rbac.yaml]: Why do we need such extensive permission for metadata-agent? Do we actually need more than pods and nodes?

File: cluster/addons/metadata-agent/stackdriver/metadata-agent.yaml [hunk 11]
```diff
@@ -45,27 +29,22 @@ spec:
...
+            port: 8000
+          initialDelaySeconds: 30
+          periodSeconds: 60
+          timeoutSeconds: 5
           failureThreshold: 1
           successThreshold: 1
-        volumeMounts:
-          - name: metadata-agent-config-volume
-            mountPath: /etc/config
-        command:
```
Comment (loburm) [inline cluster/addons/metadata-agent/stackdriver/metadata-agent.yaml]: Does it mean that new image has ENTRYPOINT specified?

File: cluster/gce/gci/configure-helper.sh [hunk 12]
```diff
@@ -2237,7 +2237,7 @@ function setup-fluentd {
     fluentd_gcp_configmap_name="fluentd-gcp-config-old"
   fi
   sed -i -e "s@{{ fluentd_gcp_configmap_name }}@${fluentd_gcp_configmap_name}@g" "${fluentd_gcp_yaml}"
-  fluentd_gcp_version="${FLUENTD_GCP_VERSION:-0.2-1.5.30-1-k8s}"
+  fluentd_gcp_version="${FLUENTD_GCP_VERSION:-0.3-1.5.34-1-k8s-1}"
```
Comment (loburm) [inline cluster/gce/gci/configure-helper.sh]: I think we have a bug in updating version. We have a huge inconsistency in labels in fluentd yaml file. Could you ensure that all labels match to the fluentd_gcp_version variable?

**Review by loburm — CHANGES_REQUESTED** [review]

File: cluster/addons/metadata-agent/stackdriver/metadata-agent-rbac.yaml (hunk 10 above)
Comment (igorpeshansky) [reply cluster/addons/metadata-agent/stackdriver/metadata-agent-rbac.yaml]: The metadata agent needs to follow the owner references to find the top-level owner for a given pod. With custom object types and operators, the owner chain may contain anything. Restricting this access will result in incomplete data.

File: cluster/addons/metadata-agent/stackdriver/metadata-agent.yaml (hunk 11 above)
Comment (igorpeshansky) [reply cluster/addons/metadata-agent/stackdriver/metadata-agent.yaml]: Yes, it does.

**Comment by igorpeshansky** [comment]
@foxish, PTAL.

File: cluster/gce/gci/configure-helper.sh (hunk 12 above)
Comment (x13n) [reply cluster/gce/gci/configure-helper.sh]: Yeah, please also bump version in cluster/addons/fluentd-gcp/fluentd-gcp-ds.yaml (3 places - DS name, DS label, pod template label).
Ideally this should be also controlled via a variable, like image version, but today it's not.

Comment (bmoyles0117) [reply cluster/gce/gci/configure-helper.sh]: Done

**Comment by MaciekPytel** [comment]
/approve 
For cluster/gce.

File: cluster/gce/config-default.sh [hunk 13]
```diff
@@ -404,6 +404,8 @@ if [[ -n "${LOGROTATE_MAX_SIZE:-}" ]]; then
 fi
 
 # Fluentd requirements
+# YAML exists to trigger a configuration refresh when changes are made.
+FLUENTD_GCP_YAML_VERSION="3.1.0"
```
[the remark this answers is not in the document]
Comment (bmoyles0117) [reply cluster/gce/config-default.sh]: Done
